### PR TITLE
Fixed backfill interference with scheduler

### PR DESCRIPTION
**Pull request by houqp**
The error message included in the commit should be fairly self-explanatory :)

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by houqp** [comment]
cc @dimonchik-suvorov

**Comment by potiuk** [comment]
Why do you change states to NONE ? Were the tests wrong to create dagruns in RUNNING state  in those tests.

Can we also have a backfill test covering the warning?

**Comment by houqp** [comment]
> Why do you change states to NONE ? Were the tests wrong to create dagruns in RUNNING state in those tests.

Yeah, we think it's a bug to allow running backfill against an active dag run managed by the scheduler. This was allowed in the current airflow main branch and this PR prevents it.

I will add a test for it today 👍

**Comment by houqp** [comment]
@potiuk test added.

File: tests/jobs/test_backfill_job.py [hunk 1]
```diff
@@ -581,11 +581,41 @@ def test_backfill_run_rescheduled(self, dag_maker):
...
+
+        job = BackfillJob(
+            dag=dag,
+            executor=executor,
+            start_date=DEFAULT_DATE,
+            end_date=DEFAULT_DATE + datetime.timedelta(days=2),
+        )
+        with patch.object(job.log, 'error') as mock_error:
+            job.run()
+            mock_error.assert_called()
```
Comment (potiuk) [inline tests/jobs/test_backfill_job.py]: NIT: I'd also do if assert "Backfill cannot be created" in the error message. It just prevents the case that there is **another** error generated during the test.

**Review by potiuk — APPROVED** [review]
Looks cool @houqp ! - just one NIT about asserting error message.

File: airflow/jobs/backfill_job.py [hunk 2]
```diff
@@ -782,6 +784,28 @@ def _execute(self, session=None):
...
+                self.log.error(
+                    "Backfill cannot be created for DagRun %s in %s, as there's already %s in a RUNNING "
+                    "state. Changing DagRun into BACKFILL would cause scheduler to lose track of executing "
+                    "tasks. Not changing DagRun type into BACKFILL, and trying insert another DagRun into "
+                    "database would cause database constraint violation for dag_id + execution_date "
+                    "combination. Please adjust backfill dates or wait for this DagRun to finish.",
+                    run.dag_id,
+                    run.execution_date.strftime("%Y-%m-%dT%H:%M:%S"),
+                    run.run_type,
+                )
```
Comment (uranusjr) [inline airflow/jobs/backfill_job.py]: Should we only log one error for all runs, or must we log one for each?

File: airflow/jobs/backfill_job.py [hunk 3]
```diff
@@ -782,6 +784,28 @@ def _execute(self, session=None):
                 return
             dagrun_infos = [DagRunInfo.interval(dagrun_start_date, dagrun_end_date)]
 
+        dag_with_subdags_ids = [d.dag_id for d in self._get_dag_with_subdags()]
+        running_dagruns = DagRun.find(
+            dag_id=dag_with_subdags_ids,
+            execution_start_date=self.bf_start_date,
+            execution_end_date=self.bf_end_date,
+            no_backfills=True,
+            state=State.RUNNING,
```
Comment (uranusjr) [inline airflow/jobs/backfill_job.py]: Please prefer `DagRunState`. Same applies for other changes. (For `State.NONE`, simply use `None` instead.)

Comment (houqp) [reply airflow/jobs/backfill_job.py]: updated

File: airflow/jobs/backfill_job.py (hunk 2 above)
Comment (houqp) [reply airflow/jobs/backfill_job.py]: I pushed a change to move majority repeated of the error log into a single log entry. Left the dynamic part to remain per entry so it still maintains the dag_id, execution_date and run_type mapping.

File: airflow/jobs/backfill_job.py [hunk 4]
```diff
@@ -784,6 +786,31 @@ def _execute(self, session=None):
...
+            state=DagRunState.RUNNING,
+        )
+
+        if running_dagruns:
+            for run in running_dagruns:
+                self.log.error(
+                    "Backfill cannot be created for DagRun %s in %s, as there's already %s in a RUNNING "
+                    "state.",
+                    run.dag_id,
+                    run.execution_date.strftime("%Y-%m-%dT%H:%M:%S"),
```
Comment (uranusjr) [inline airflow/jobs/backfill_job.py:805]: Why not use `run.run_id` here?

Comment (uranusjr) [inline airflow/jobs/backfill_job.py:805]: Why not use `run.run_id` here?

File: airflow/jobs/backfill_job.py [hunk 5]
```diff
@@ -290,7 +290,6 @@ def _get_dag_run(self, dagrun_info: DagRunInfo, dag: DAG, session: Session = Non
         current_active_dag_count = dag.get_num_active_runs(external_trigger=False)
 
         # check if we are scheduling on top of a already existing dag_run
-        # we could find a "scheduled" run instead of a "backfill"
```
Comment (uranusjr) [inline airflow/jobs/backfill_job.py]: Is this removed intentionally and why?

Comment (uranusjr) [inline airflow/jobs/backfill_job.py]: Is this removed intentionally and why?

File: tests/conftest.py [hunk 6]
```diff
@@ -500,11 +500,12 @@ def create_dagrun(self, **kwargs):
 
             dag = self.dag
             kwargs = {
-                "state": State.RUNNING,
                 "start_date": self.start_date,
                 "session": self.session,
                 **kwargs,
             }
+            if "state" not in kwargs:
+                kwargs["state"] = State.RUNNING
```
Comment (uranusjr) [inline tests/conftest.py]: This should not be needed, the dict unpacking syntax already guarantees `"state"` in `**kwargs` takes precedence over the one above it.

Comment (uranusjr) [inline tests/conftest.py]: This should not be needed, the dict unpacking syntax already guarantees `"state"` in `**kwargs` takes precedence over the one above it.

File: tests/jobs/test_backfill_job.py [hunk 7]
```diff
@@ -608,11 +609,44 @@ def test_backfill_override_conf(self, dag_maker):
...
+
+        executor = MockExecutor()
+
+        job = BackfillJob(
+            dag=dag,
+            executor=executor,
+            start_date=DEFAULT_DATE,
+            end_date=DEFAULT_DATE + datetime.timedelta(days=2),
+        )
+        with patch.object(job.log, 'error') as mock_error:
```
Comment (uranusjr) [inline tests/jobs/test_backfill_job.py]: Don't need to mock this, you can use pytest’s `caplog` fixture.

Comment (uranusjr) [inline tests/jobs/test_backfill_job.py]: Don't need to mock this, you can use pytest’s `caplog` fixture.

**Comment by uranusjr** [comment]
A few minor comments on implementation, the general logic looks right to me.

**Comment by uranusjr** [comment]
A few minor comments on implementation, the general logic looks right to me.

File: tests/jobs/test_backfill_job.py (hunk 7 above)
Comment (houqp) [reply tests/jobs/test_backfill_job.py]: TIL, thanks for the tip!

File: airflow/jobs/backfill_job.py (hunk 5 above)
Comment (houqp) [reply airflow/jobs/backfill_job.py]: good catch, removed by mistake.

File: tests/jobs/test_scheduler_job.py [hunk 8]
```diff
@@ -1760,7 +1760,7 @@ def evaluate_dagrun(
         dr = dag.create_dagrun(
             run_type=DagRunType.SCHEDULED,
             execution_date=dagrun_info.logical_date,
-            state=State.RUNNING,
```
Comment (ashb) [inline tests/jobs/test_scheduler_job.py:1763]: Why did you need to change these tests?

Comment (houqp) [reply tests/jobs/test_scheduler_job.py:1763]: because these tests created a scheduled dagrun in running state and this fix prevents a backfill run from overriding a scheduled DAG run.

Comment (ashb) [reply tests/jobs/test_scheduler_job.py:1763]: Ohh, yes, the scheduler tests that use `dag.run` which actually uses BackfillJob. We should fix that at some point.

**Review by ashb — APPROVED** [review]
